File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.py
from collections import deque
# bfs => queue
def solution(maps):
    answer = 0
    
    n=len(maps)
    m=len(maps[0])
    
    dx = [0, 0, -1, 1]
    dy = [-1, 1, 0, 0]
    
    queue = deque()
    queue.append((0,0,1))
    
    visited = [[False]*m for _ in range(n)]
    visited[0][0]=True
    
    while queue:
        x,y,dist = queue.popleft()
        
        if x==n-1 and y==m-1:
            return dist
        
        for i in range(4):
            nx = x + dx[i]
            ny = y + dy[i]
            
            if 0<= nx< n and 0<=ny<m and not visited[nx][ny] and maps[nx][ny] ==1:
                visited[nx][ny]=True
                queue.append((nx, ny, dist +1))
    
    
    
    return -1

File: 프로그래머스/2/1844. 게임 맵 최단거리/게임 맵 최단거리.cpp
#include<vector>
#include<queue>

using namespace std;

int solution(vector<vector<int>> maps)
{
    int answer = 0;
    int R = maps.size();
    int C = maps[0].size();
    
    vector<vector<int>> dist(R, vector<int>(C, -1));
    queue<pair<int,int>> q;
    
    int dr[4] = {-1,1,0,0};
    int dc[4] = {0,0,-1,1};
    
    q.push({0,0});
    dist[0][0]=1;
    
    while(!q.empty()){
        auto [r, c] = q.front();
        q.pop();
        
        for(int d=0; d<4; d++){
            int nr = r+dr[d];
            int nc = c+dc[d];
            
            if(nr>=0 && nr<R && nc>=0 && nc<C && maps[nr][nc]==1 && dist[nr][nc]==-1){
                dist[nr][nc] = dist[r][c]+1;
                q.push({nr, nc});
            }
        }
    }    
    return dist[R-1][C-1];
}